File: Leetcode/L_Longest_Common_Prefix.py
# Longest Commmon Prefix

from collections import Counter


class Solution:
    def longestCommonPrefix(self, strs):
        if len(strs) == 0:
            return ""
        minlen = len(strs[0])
        for i in range(len(strs)):
            minlen = min(len(strs[i]), minlen)
        lcp = ""
        i = 0
        while i < minlen:
            char = strs[0][i]
            for j in range(1, len(strs)):
                if strs[j][i] != char:
                    return lcp
            lcp = lcp + char
            i += 1
        return lcp


X = Solution()
print(X.longestCommonPrefix(["flower", "flow", "flight"]))

File: Leetcode Js/L_Longest_Common_Prefix.js
// console.log("Hello World");
// Longest Common Prefix
// Important string question asked in Faang
// This question is tricky question and have 5 different approach to solve it

// Approach 1 - Word by Word
// T - O(NM)
// S - O(M)

/**
 * @param {string[]} strs
 * @return {string}
 */

var commonPrefixUtil = (str1, str2) => {
  let result = "";
  let n1 = str1.length,
    n2 = str2.length;
  for (let i = 0; i < n1; i++) {
    for (let j = 0; j < n2; j++) {
      if (str1[i] !== str2[i]) {
        break;
      }
      result = result + str1[i];
    }
  }
  return result;
};

var longestCommonPrefix = function (strs) {
  var n = strs.length;
  var prefix = strs[0];
  for (let i = 0; i < n; i++) {
    prefix = commonPrefixUtil(prefix, strs[i]);
  }
  return prefix;
};

console.log(longestCommonPrefix(["flower", "flow", "flight"]));
console.log(longestCommonPrefix(["Engineer", "Abhishek", "Verma"]));
console.log(longestCommonPrefix(["geeksforgeeks", "geeks", "geek", "geezer"]));
